File: python/leetcode22.py
class Solution:
    def generateParenthesis(self, n):
        self.n = n
        self.result = []
        self.generate(0, 0)
        return self.result

    def generate(self, left, right, s=""):
        if len(s) == self.n * 2:
            self.result.append(s)
            return 
        
        if left < self.n:
            self.generate(left+1, right, s+"(")
        
        if right < left:
            self.generate(left, right+1, s+")")
        

if __name__ == "__main__":
    print(Solution().generateParenthesis(3))

File: cpp/leetcode22.cpp
#include<vector>
#include<string>

using namespace std;

class Solution {
vector<string> arr;
public:
    vector<string> generateParenthesis(int n) {
        generate(0, 0, "", n);
        return arr;
    }

    void generate(int left, int right, string s, int n){
        if(s.size() == 2 * n) {
            arr.push_back(s);
            return;
        }
        if(left < n){
            generate(left+1, right, s+"(", n);
        }
        if (right < left){
            generate(left, right+1, s+")", n);
        }
    }
};

